feat: extend track parameters accessors

msmk0 opened the pull request:
This adds the following additional accessors for derived quantities to all track parameters:

```cpp
.fourPosition(...)
.unitDirection()
.transverseMomentum()
.absoluteMomentum()
```

Having separate access to the direction and the absolute momentum provides the quantities that are actually used e.g. by the propagators. It is also closer to what is stored in the parameters vectors and reduces unnecessary computations.

This also removes the zero-arg `.position()` access for the `SingleCurvilinearTrackParameters` since it can not be consistently implemented without requiring a default geometry context.

The parameters concepts are updated to require `.fourPosition(...)`, `.unitDirection()` and `.absoluteMomentum()` instead of `.momentum()`. All track parameters retain the `.momentum()` accessor as a convenience for users, but it is not used internally in the code base.

Fixes #152, fixes #153, and prepares for #371.

paulgessinger commented:
What about q/p? And why `absMomentum`?

paulgessinger commented on Core/include/Acts/EventData/SingleBoundTrackParameters.hpp:
```diff
@@ -142,13 +142,31 @@ class SingleBoundTrackParameters {
+    const Vector2D loc(get<eBoundLoc0>(), get<eBoundLoc1>());
+    const Vector3D dir =
+        makeDirectionUnitFromPhiTheta(get<eBoundPhi>(), get<eBoundTheta>());
+    Vector4D pos4 = Vector4D::Zero();
```
You don't need the zero init I think.

msmk0 replied:
I have had uninitialized warnings in the past and it is good practice to always initialize all variables to a known value. I hope that the compiler optimizes the initial zeroing away anyways.

paulgessinger replied:
I wouldn't consider it good practice, but ok. I'm also not sure this will be optimized away, so there might be a (admittedly small) performance impact.

msmk0 replied:
Removed.

paulgessinger commented on Core/include/Acts/EventData/SingleBoundTrackParameters.hpp:
```diff
@@ -158,19 +176,18 @@ class SingleBoundTrackParameters {
-    mom *= std::abs(1 / get<eBoundQOverP>());
-    return mom;
+  /// Absolute transverse momentum.
+  Scalar transverseMomentum() const {
```
Why rename this? `pT` is pretty much universal.

msmk0 replied:
The idea was to be consistent with the rest of the accessors. All of them are written out, e.g. `position` and `charge`. Thus, `transverseMomentum` and `absMomentum`. This is also used by `ActsFatras::Particle`.

Another option could be `.pt()` and `.p()`?

paulgessinger replied:
Well if you write this out it should be `absoluteMomentum` as well. I think these names are all longer than they need to be.

msmk0 replied:
Rename to `absoluteMomentum` for consistency.

paulgessinger commented on Core/include/Acts/EventData/SingleBoundTrackParameters.hpp:
```diff
@@ -158,19 +176,18 @@ class SingleBoundTrackParameters {
   Scalar time() const { return get<eBoundTime>(); }
 
-  /// Access the direction pseudo-rapidity.
-  Scalar eta() const { return -std::log(std::tan(get<eBoundTheta>() / 2)); }
```
Do we not need this anywhere?

msmk0 replied:
No, this is not used anywhere in the code. All use cases are served by `VectorHelpers::eta(...)` already.

paulgessinger replied:
Ok.

paulgessinger commented on Core/include/Acts/EventData/SingleBoundTrackParameters.hpp:
```diff
@@ -142,13 +142,31 @@ class SingleBoundTrackParameters {
+  /// surface to globalcoordinates. This requires a geometry context to select
+  /// the appropriate transformation and might be a computationally expensive
+  /// operation.
+  Vector4D position4(const GeometryContext& geoCtx) const {
```
Following your naming, why not `fourPosition`?

msmk0 replied:
That could work, too. I prefer `position4` because it is a bit shorter and because it is consistent with existing code in `ActsFatras::{Hit,Particle}`.

paulgessinger commented on Core/include/Acts/EventData/SingleCurvilinearTrackParameters.hpp:
```diff
@@ -68,14 +68,21 @@ class SingleCurvilinearTrackParameters
+  ///
+  /// The surface is owned by the parameters object and thus is independent
+  /// from the geometry context.
+  Vector4D position4() const { return Base::position4(GeometryContext()); }
```
I don't think we should default construct the geo context anywhere in our code other than the tests.

msmk0 replied:
I agree in principle, but there is no other way here. The curvilinear parameters can be defined anywhere in space and are not restricted to the existing geometry. They own there surface and accessing the position is independent from the geometry context. But, all surface transformations require a context. So it is either that, or to not provide `.position{4}()` at all and require an external context also for the curvilinear parameters.

paulgessinger replied:
You can reimplement the getter, and I think you should in this case.

msmk0 replied:
It still needs to use `Surface::localToGlobal` which always requires a geometry context. Rewritting the getter will end up with exactly the same code as before just with an addditional
```
GeometryContext geoCtx;
```
declaration.

paulgessinger replied:
Hm, I'm not happy about this, but ok.

paulgessinger replied:
On second thought, I don't we do (or event should) require `GeometryContext` to be default-constructible outside of our test code, where we plug in an example context.

paulgessinger replied:
Also, in curvilinear parameters we know the contained surface is a plane surface, so the transform of that is enough. Therefore, just reimplement the `Base::position` here, don't use `localToGlobal` but put a manual transform multiplication in.

msmk0 replied:
> Also, in curvilinear parameters we know the contained surface is a plane surface, so the transform of that is enough. Therefore, just reimplement the `Base::position` here, don't use `localToGlobal` but put a manual transform multiplication in.

I know, but that does not help us. One of the main points of the parameters clean up was to reduce the number of variants that we need to support; thus, `CurvilinearTrackParameters` are now just `BoundTrackParameters` with a fixed internal surface. But that also means that we need to support `.referenceSurface()` which must return a `Surface` reference. That means that we must store the `PlaneSurface` and not just the transformation. We could `dynamic_cast<....>` the surface interface to `PlaneSurface` but the `PlaneSurface` does not provide access to the internal transformation **without geometry context** either. As far as I understand it, it can not provide this since the underlying `Surface` could always be based on an underlying detector element and thus **must** be accessed via a geometry context.

So i really do not see how this can be implemented without a default constructed geometry context. I agree with you that we can not assume that the geometry context can be default constructed. My suggestion would then be to remove the geometry context independent position-like accessors altogether for now.

paulgessinger replied:
True. We could add a getter `defaultTransform` to surface that acts as just a dumb getter for the transform. That would be preferrable to giving meaning to default constructed `GeometryContext` in my opinion.

Otherwise, I think removing the zero-arg version here would be the way to go.

msmk0 replied:
The zero-arg versions are removed (as mentioned below).

paulgessinger commented on Core/include/Acts/EventData/SingleFreeTrackParameters.hpp:
```diff
@@ -101,31 +101,48 @@ class SingleFreeTrackParameters {
-  /// @return 3D vector with global position
+  /// Space-time position four-vector.
+  Vector4D position4() const {
+    Vector4D pos4 = Vector4D::Zero();
```
Init not needed.

msmk0 replied:
Removed.

paulgessinger commented on Tests/UnitTests/Core/EventData/FreeTrackParametersTests.cpp:
```diff
@@ -20,13 +20,20 @@ namespace Test {
+  double q = 1.;
 
   FreeVector params;
-  params << pos.x(), pos.y(), pos.z(), t, dir.x(), dir.y(), dir.z(), qop;
```
I understand why you do this, but it's painful.

msmk0 replied:
I do not disagree, but it is the only way if we want to allow configurable parameter ordering.

paulgessinger commented:
Also why `directionUnit` rather than the existing `unitDirection`? The latter (and current) ordering is more in line with the other name ordering you add here.

msmk0 commented:
> Also why `directionUnit` rather than the existing `unitDirection`? The latter (and current) ordering is more in line with the other name ordering you add here.

Yes, you are right. I am not sure exactly why I changed the ordering here. I will fix that.

msmk0 commented:
> > Also why `directionUnit` rather than the existing `unitDirection`? The latter (and current) ordering is more in line with the other name ordering you add here.
> 
> Yes, you are right. I am not sure exactly why I changed the ordering here. I will fix that.

I switched to `unitDirection` as requested.

paulgessinger commented on Core/include/Acts/EventData/TrackParametersConcept.hpp:
```diff
@@ -26,24 +26,32 @@ using TypeParametersVector = typename T::ParametersVector;
 using ReturnTypeCovariance = decltype(std::declval<T>().covariance());
 template <typename T>
+using ReturnTypePosition4FromContext =
+    decltype(std::declval<T>().position4(GeometryContext()));
```
This should use `std::declval`. `GeometryContext` is a typedef over `std::any`, and while that is actually default-constructible, we should not require it. Also, there's no implicit meaning that *empty* `GeometryContext` means *nominal alignment*.

msmk0 replied:
Fixed.

paulgessinger commented on Core/include/Acts/EventData/TrackParametersConcept.hpp:
```diff
@@ -26,24 +26,32 @@ using TypeParametersVector = typename T::ParametersVector;
+using ReturnTypePosition4 = decltype(std::declval<T>().position4());
+template <typename T>
 using ReturnTypePositionFromContext =
     decltype(std::declval<T>().position(GeometryContext()));
```
Same thing here, even though this is not part of the changeset.

msmk0 replied:
Fixed.

paulgessinger commented on Core/include/Acts/EventData/SingleCurvilinearTrackParameters.hpp:
```diff
@@ -68,14 +68,21 @@ class SingleCurvilinearTrackParameters
   ///
   /// The surface is owned by the parameters object and thus is independent
   /// from the geometry context.
   Vector3D position() const { return Base::position(GeometryContext()); }
```
Same here, even though this is not in the changeset.

paulgessinger reviewed (changes requested):
I think we'll have to work around default-constructing `GeometryContext`. Other than that, I'm ok with this.

msmk0 commented:
@paulgessinger You are right, `fourPosition` would be more consistent. I will rename the accesors and switch to `declval`

msmk0 commented:
@paulgessinger I update the PR as discussed. The context-independent position access for curvilinear parameters is removed. Trying to solve this in a consistent manner will require some more work and it is not really used at the moment. Adding this would be a backwards-compatible change anyway so we can easily do it later.

paulgessinger reviewed (approved):
Ok I think.

paulgessinger commented:
Can you update the PR description @msmk0?

msmk0 commented:
The PR description is now up-to-date.

asalzburger commented:
Nice!
